File: c/problem_49.c
#include "problem_49.h"
#include <stdio.h>
#include <assert.h>

long long max_subarray_sum(long long *arr, int n) {
    long long best = 0, cur = 0;
    for (int i = 0; i < n; i++) {
        cur += arr[i];
        if (cur > best) best = cur;
        if (cur < 0) cur = 0;
    }
    return best;
}

void problem49(void) {
    printf("Problem 49:\n");
    long long a[] = {34,-50,42,14,-5,86};
    assert(max_subarray_sum(a, 6) == 137);
    long long b[] = {-5,-1,-8,-9};
    assert(max_subarray_sum(b, 4) == 0);
    long long c[] = {1,2,3};
    assert(max_subarray_sum(c, 3) == 6);
    printf("All Tests Passed\n");
}

File: lua/problem_49.lua
local M = {}

function M.max_subarray_sum(arr)
    local best, cur = 0, 0
    for _, v in ipairs(arr) do
        cur = math.max(0, cur + v)
        best = math.max(best, cur)
    end
    return best
end

function M.problem49()
    print("Problem 49:")
    assert(M.max_subarray_sum({34,-50,42,14,-5,86}) == 137)
    assert(M.max_subarray_sum({-5,-1,-8,-9})         == 0)
    assert(M.max_subarray_sum({1,2,3})               == 6)
    print("All Tests Passed")
end

return M
